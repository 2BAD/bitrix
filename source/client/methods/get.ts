import { GotInstance, GotJSONFn } from 'got'
import { stringify as toQuery } from 'qs'
import { GetParams, GetPayload, GettableMethod } from '../types'

export type Get = <P>(method: GettableMethod, params?: GetParams) => Promise<GetPayload<P>>

/**
 * Dispatches a request with specified method and params.
 * Should be used for getting single items only, otherwise payload will be wrongly typed.
 */
export default (got: GotInstance<GotJSONFn>): Get => {
  const get: Get = <P>(method: GettableMethod, params?: GetParams): Promise<GetPayload<P>> =>
    got.get(method, { query: toQuery(params) })
        .then(({ body }) => body as GetPayload<P>)

  return get
}
