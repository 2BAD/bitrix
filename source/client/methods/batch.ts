import { GotJSONFn } from 'got'
import chunk from 'lodash.chunk'
import fromPairs from 'lodash.frompairs'
import { stringify as toQuery } from 'qs'
import { Commands } from '../../commands'
import { Method, MethodPayloadType } from '../../methods'
import { BatchPayload } from '../../payloads'
import isArray from '../../utils/isArray'
import { handlePayload } from './call'

export const MAX_COMMANDS_PER_BATCH = 50

/**
 * Split any number of commands into the chunks with max commands per batch in each chunk
 */
export const chunkCommands = <C extends Commands>(
  commands: C,
  size: number = MAX_COMMANDS_PER_BATCH
): ReadonlyArray<C> => {
  const commandsIsArray = isArray(commands)
  const chunks = chunk(Object.entries(commands), size)

  // @todo Generated return type isn't very accurate since it says that it will make array of <C>, while
  //       in fact <C> will be distributed among chunks. That gives wrong hints with named commands and arrays as const.
  return chunks.map(
    (c) => commandsIsArray
      // @todo Figure out how to avoid unsafe casting to <C> here
      //       Without `<C>` generic it won't work with complex functions like `batch`
      ? c.map(([_key, command]) => command) as unknown as C
      : fromPairs(c) as C
  )
}

/**
 * Prepares list of commands to be used in requests to the Bitrix
 * @note We could avoid that function... if only Bitrix API would support posting of the batch commands as
 *       plain objects, like it does with other methods. Instead, it should be a dict or array of string queries
 */
export const prepareCommandsQueries = <C extends Commands, R = { [K in keyof C]: string }>(commands: C): R =>
  Object.entries(commands).reduce((calls, [name, { method, params }]) => {
    const stringifiedParams = params ? `?${toQuery(params)}` : ''

    return {
      ...calls,
      [`cmd[${name}]`]: `${method}${stringifiedParams}`
    }
  // tslint:disable-next-line: no-object-literal-type-assertion
  }, {} as R)

/**
 * Merges list of batch responses into a single batch
 * @todo Generics inference is complicated here and might be not super accurate
 */
export const mergeBatchPayloads = <
  B extends BatchPayload<any>,
  P = B extends BatchPayload<infer U> ? U : never
>(payloads: ReadonlyArray<B>): BatchPayload<P> => {
  const merge = <T extends S, S extends readonly any[] | Record<string, any>>(a: T, b: S): T =>
    isArray(a) && isArray(b)
      // @todo Gotta find way to avoid unsafe casting here
      ? [...a, ...b] as unknown as T
      : ({ ...a, ...b })

  return payloads.reduce((merged, payload) => ({
    result: {
      result: merge(merged.result.result || [], payload.result.result),
      result_error: merge(merged.result.result_error || [], payload.result.result_error),
      result_next: merge(merged.result.result_next || [], payload.result.result_next),
      result_time: merge(merged.result.result_time || [], payload.result.result_time),
      result_total: merge(merged.result.result_total || [], payload.result.result_total)
    },
    time: { ...merged.time, ...payload.time }
  // tslint:disable-next-line: no-object-literal-type-assertion
  }), { result: {}, time: {} } as BatchPayload<P>)
}

export type CommandsPayloads<
  C extends Commands,
  CM = {
    [K in keyof C]: C[K] extends { readonly method: infer M } ? M : never
  }
  > = {
  [K in keyof CM]: MethodPayloadType<CM[K] extends Method ? CM[K] : never>
}

// @todo Using `CommandsWithRequiredParamsTypes` and `Batch2` we could ensure that only supported commands
//       are supplied to the `batch`, but it results in unknown error when assigning `Batch2` to `batch` constant.

// export type CommandsWithRequiredParamsTypes<
//   C extends Commands,
//   CC = {
//     [K in keyof C]: C[K] extends Command ? {
//       method: C[K]['method'], params: MethodParams<C[K]['method']>
//     } : Command
//   },
// > = CC

// export type Batch2 = <
//   C extends Commands & { [K in keyof CC]: CC[K] },
//   CC = CommandsWithRequiredParamsTypes<C>,
//   P = CommandsPayloads<C>
// >(commands: C, commandsPerRequest?: number) => Promise<BatchPayload<P>>

export type Batch = <
  C extends Commands,
  P = CommandsPayloads<C>
>(commands: C, commandsPerRequest?: number) => Promise<BatchPayload<P>>

interface Dependencies {
  readonly get: GotJSONFn
}

/**
 * Dispatches a batch request with specified commands. Will fill figure out payload type based on the Methods.
 * Supports unlimited number of commands. If they do exceed max amount of commands per batch, will dispatch
 * multiple request and merge result into single batch payload
 * @todo For now due to issues with types it won't check `params` of command, so any Method can be used
 *       with any (even wrong) params
 */
export default ({ get }: Dependencies): Batch => {
  const batch: Batch = async <
    C extends Commands,
    P = CommandsPayloads<C>
  >(
    commands: C,
    commandsPerRequest: number = MAX_COMMANDS_PER_BATCH
  ): Promise<BatchPayload<P>> => {
    const call = (c: C) =>
      get(Method.BATCH, { query: prepareCommandsQueries(c) })
        .then(({ body }) => body as BatchPayload<P>)

    const calls = chunkCommands(commands, commandsPerRequest)
      .map(call)

    return Promise.all(calls)
      .then((chunkedPayloads) => {
        const payloads = mergeBatchPayloads(chunkedPayloads)
        return handlePayload(payloads)
      })
  }

  return batch
}
