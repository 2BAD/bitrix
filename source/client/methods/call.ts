import got from 'got'
import { stringify as toQuery } from 'qs'
import { Method, MethodParams, MethodPayload } from '../../methods'
import { BatchPayload, ListPayload, Payload } from '../../payloads'
import isArray from '../../utils/isArray'

/**
 * Checks wether payload have any errors and if it does â€” throws them
 * Bitrix payload types do not provide discriminators, so we're forced to type cast them
 */
export const handlePayload = <P extends Payload<unknown>>(payload: P): P => {
  if ((payload as ListPayload<unknown>).error) {
    throw new Error(
      `[call] failed to get the resource: ${(payload as ListPayload<unknown>).error ?? ''}.`
    )
  }

  if ((payload as BatchPayload<unknown>).result && (payload as BatchPayload<unknown>).result.result_error) {
    const resultErrors = (payload as BatchPayload<unknown>).result.result_error
    const errors = isArray(resultErrors) ? resultErrors : Object.values(resultErrors)

    if (errors.length > 0) {
      // @todo We can give better formatting to display errored commands. But it's not important for now
      throw new Error(`[batch] failed to process. Received errors in ${errors.length} commands:\n${errors.join('\n')}`)
    }
  }

  return payload
}

export type Call = <M extends Method>(method: M, params: MethodParams<M>) => Promise<MethodPayload<M>>

type Dependencies = {
  readonly get: typeof got.get
}

/**
 * Dispatches a request with specified method and params. Will fill figure out payload type based on the Method
 */
export default ({ get }: Dependencies): Call => {
  const call: Call = <M extends Method>(method: M, params: MethodParams<M>): Promise<MethodPayload<M>> =>
    get<MethodPayload<M>>(method, { searchParams: toQuery(params) })
      .then(({ body }) => handlePayload(body))

  return call
}
