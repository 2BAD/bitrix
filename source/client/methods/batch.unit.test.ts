import got from 'got'
import range from 'lodash.range'
import nock from 'nock'
import { describe, expect, it } from 'vitest'
import { Method } from '../../methods.js'
import Batch, {
  chunkCommands,
  MAX_COMMANDS_PER_BATCH,
  mergeBatchPayloads,
  prepareCommandsQueries
} from './batch.js'

describe('Client `chunkCommands` method', () => {
  it('should chunk named commands', () => {
    const chunkSize = 2
    const commands = {
      a: { method: Method.CRM_DEAL_GET },
      b: { method: Method.CRM_LEAD_GET },
      c: { method: Method.CRM_DEAL_LIST },
      d: { method: Method.CRM_LEAD_LIST },
      e: { method: Method.CRM_STATUS_LIST }
    }

    expect(chunkCommands(commands, chunkSize)).toMatchSnapshot()
  })

  it('should chunk array of commands', () => {
    const chunkSize = 2
    const commands = [
      { method: Method.CRM_DEAL_GET },
      { method: Method.CRM_LEAD_GET },
      { method: Method.CRM_DEAL_LIST },
      { method: Method.CRM_LEAD_LIST },
      { method: Method.CRM_STATUS_LIST }
    ] as const

    expect(chunkCommands(commands, chunkSize)).toMatchSnapshot()
  })

  it('should by default chunk with a size of `MAX_COMMANDS_PER_BATCH`', () => {
    const commandsSets = 2
    const commands = range(0, MAX_COMMANDS_PER_BATCH * commandsSets).map(() => ({
      method: Method.CRM_DEAL_GET
    }))

    const chunked = chunkCommands(commands)

    expect(chunked.length).toBe(commandsSets)
    // @ts-ignore we have asserted it length
    expect(chunked[0].length).toBe(MAX_COMMANDS_PER_BATCH)
  })
})

describe('Client `prepareCommandsQueries` method', () => {
  it('should transform dict of the commands into the query object', () => {
    const testDealId = 11111

    const commands = {
      one: {
        method: Method.CRM_DEAL_GET,
        params: { ID: testDealId }
      },
      two: {
        method: Method.CRM_DEAL_LIST,
        params: {
          filter: { '>PROBABILITY': 50 },
          order: { STAGE_ID: 'ASC' },
          select: ['ID', 'TITLE'],
          start: 200
        }
      }
    }

    expect(prepareCommandsQueries(commands)).toMatchSnapshot()
  })

  it('should work with numbered commands', () => {
    const commands = {
      0: { method: Method.CRM_DEAL_GET },
      1: { method: Method.CRM_DEAL_LIST }
    }

    expect(prepareCommandsQueries(commands)).toMatchSnapshot()
  })

  it('should work with array of commands', () => {
    const commands = [{ method: Method.CRM_DEAL_GET }, { method: Method.CRM_DEAL_LIST }] as const

    expect(prepareCommandsQueries(commands)).toMatchSnapshot()
  })

  it('should return empty query object when no commands provided', () => {
    expect(prepareCommandsQueries({})).toMatchSnapshot()
  })
})

describe('Client `mergeBatchPayloads` method', () => {
  it('should merge named payloads', () => {
    const batch1 = {
      result: {
        result: {
          a: [{ ID: '1' }, { ID: '2' }],
          b: [{ ID: '3' }, { ID: '4' }]
        },
        result_error: [],
        result_total: { a: 8, b: 8 },
        result_next: { a: 2, b: 4 },
        result_time: {
          a: {
            start: 1567196891.008149,
            finish: 1567196891.022234,
            duration: 0.014085054397583008,
            processing: 0.013998985290527344,
            date_start: '2019-08-30T23:28:11+03:00',
            date_finish: '2019-08-30T23:28:11+03:00'
          },
          b: {
            start: 1567196891.022316,
            finish: 1567196891.03225,
            duration: 0.009933948516845703,
            processing: 0.009846210479736328,
            date_start: '2019-08-30T23:28:11+03:00',
            date_finish: '2019-08-30T23:28:11+03:00'
          }
        }
      },
      time: {
        start: 1567196890.959017,
        finish: 1567196891.223739,
        duration: 0.2647218704223633,
        processing: 0.21567082405090332,
        date_start: '2019-08-30T23:28:10+03:00',
        date_finish: '2019-08-30T23:28:11+03:00'
      }
    }

    const batch2 = {
      result: {
        result: {
          c: [{ ID: '5' }, { ID: '6' }],
          d: [{ ID: '7' }, { ID: '8' }]
        },
        result_error: [],
        result_total: { c: 8, d: 8 },
        result_next: { c: 6, d: 8 },
        result_time: {
          c: {
            start: 1567196891.032315,
            finish: 1567196891.035297,
            duration: 0.002981901168823242,
            processing: 0.002897024154663086,
            date_start: '2019-08-30T23:28:11+03:00',
            date_finish: '2019-08-30T23:28:11+03:00'
          },
          d: {
            start: 1567196891.03536,
            finish: 1567196891.039936,
            duration: 0.004575967788696289,
            processing: 0.00450897216796875,
            date_start: '2019-08-30T23:28:11+03:00',
            date_finish: '2019-08-30T23:28:11+03:00'
          }
        }
      },
      time: {
        start: 1567196890.959517,
        finish: 1567196891.225739,
        duration: 0.2647218704223683,
        processing: 0.21567082405090832,
        date_start: '2019-08-30T23:28:12+03:00',
        date_finish: '2019-08-30T23:28:13+03:00'
      }
    }

    expect(mergeBatchPayloads([batch1, batch2])).toMatchSnapshot()
  })

  it('should merge array payloads', () => {
    const batch1 = {
      result: {
        result: [
          [{ ID: '1' }, { ID: '2' }],
          [{ ID: '3' }, { ID: '4' }]
        ],
        result_error: [],
        result_total: [8, 8],
        result_next: [2, 4],
        result_time: [
          {
            start: 1567196891.008149,
            finish: 1567196891.022234,
            duration: 0.014085054397583008,
            processing: 0.013998985290527344,
            date_start: '2019-08-30T23:28:11+03:00',
            date_finish: '2019-08-30T23:28:11+03:00'
          },
          {
            start: 1567196891.022316,
            finish: 1567196891.03225,
            duration: 0.009933948516845703,
            processing: 0.009846210479736328,
            date_start: '2019-08-30T23:28:11+03:00',
            date_finish: '2019-08-30T23:28:11+03:00'
          }
        ]
      },
      time: {
        start: 1567196890.959017,
        finish: 1567196891.223739,
        duration: 0.2647218704223633,
        processing: 0.21567082405090332,
        date_start: '2019-08-30T23:28:10+03:00',
        date_finish: '2019-08-30T23:28:11+03:00'
      }
    }

    const batch2 = {
      result: {
        result: [
          [{ ID: '5' }, { ID: '6' }],
          [{ ID: '7' }, { ID: '8' }]
        ],
        result_error: [],
        result_total: [8, 8],
        result_next: [6, 8],
        result_time: [
          {
            start: 1567196891.032315,
            finish: 1567196891.035297,
            duration: 0.002981901168823242,
            processing: 0.002897024154663086,
            date_start: '2019-08-30T23:28:11+03:00',
            date_finish: '2019-08-30T23:28:11+03:00'
          },
          {
            start: 1567196891.03536,
            finish: 1567196891.039936,
            duration: 0.004575967788696289,
            processing: 0.00450897216796875,
            date_start: '2019-08-30T23:28:11+03:00',
            date_finish: '2019-08-30T23:28:11+03:00'
          }
        ]
      },
      time: {
        start: 1567196890.959517,
        finish: 1567196891.225739,
        duration: 0.2647218704223683,
        processing: 0.21567082405090832,
        date_start: '2019-08-30T23:28:12+03:00',
        date_finish: '2019-08-30T23:28:13+03:00'
      }
    }

    expect(mergeBatchPayloads([batch1, batch2])).toMatchSnapshot()
  })
})

const TEST_URI = 'https://test.com/rest'
const batch = Batch({
  get: got.extend({ prefixUrl: TEST_URI, responseType: 'json' }).get
})
const RESPONSE_200 = 200

describe('Client `batch` method', () => {
  it('should form a proper request', async () => {
    const payload = { result: { result: { one: 'done', two: 'done' }, result_error: [] } }
    const dealId = 999

    const commands = {
      one: { method: Method.CRM_DEAL_GET, params: { ID: dealId } },
      two: { method: Method.CRM_DEAL_LIST }
    }

    const scope = nock(TEST_URI)
      // @todo We'd want to use `query` object here as it is much more readable, but nock for some reason
      //       fails to match request when it contains `cmd[someName]`. The issue definitely connected
      //       to the `[]`, since it does not appear when only one bracket present
      .get(
        `/${Method.BATCH}?cmd%5Bone%5D=${commands.one.method}%3FID%3D${dealId}&cmd%5Btwo%5D=${commands.two.method}`
      )
      .reply(RESPONSE_200, payload)

    await batch(commands)

    expect(scope.done()).toBe(undefined)
  })

  it('should form a proper request with numbered commands', async () => {
    const payload = { result: { result: ['done1', 'done2'], result_error: [] } }
    const dealId = 999

    const commands = {
      0: { method: Method.CRM_DEAL_GET, params: { ID: dealId } },
      1: { method: Method.CRM_DEAL_LIST }
    } as const

    const scope = nock(TEST_URI)
      // @todo We'd want to use `query` object here as it is much more readable, but nock for some reason
      //       fails to match request when it contains `cmd[someName]`. The issue definitely
      //       connected to the `[]` since it does not appear when only one bracket present
      .get(
        `/${Method.BATCH}?cmd%5B0%5D=${commands[0].method}%3FID%3D${dealId}&cmd%5B1%5D=${commands[1].method}`
      )
      .reply(RESPONSE_200, payload)

    await batch(commands)

    expect(scope.done()).toBe(undefined)
  })

  it('should form a proper request with array of commands', async () => {
    const payload = { result: { result: ['done1', 'done2'], result_error: [] } }
    const dealId = 999

    const commands = [
      { method: Method.CRM_DEAL_GET, params: { ID: dealId } },
      { method: Method.CRM_DEAL_LIST }
    ] as const

    const scope = nock(TEST_URI)
      // @todo We'd want to use `query` object here as it is much more readable, but nock for some reason
      //       fails to match request when it contains `cmd[someName]`. The issue definitely
      //       connected to the `[]` since it does not appear when only one bracket present
      .get(
        `/${Method.BATCH}?cmd%5B0%5D=${commands[0].method}%3FID%3D${dealId}&cmd%5B1%5D=${commands[1].method}`
      )
      .reply(RESPONSE_200, payload)

    await batch(commands)

    expect(scope.done()).toBe(undefined)
  })

  it('should form a proper requests with more than max allowed commands per batch', async () => {
    const maxCommandsPerBatch = 1
    const payload = { result: { result: { one: 'done', two: 'done' }, result_error: [] } }
    const dealId = 999

    const commands = {
      one: { method: Method.CRM_DEAL_GET, params: { ID: dealId } },
      two: { method: Method.CRM_DEAL_LIST }
    }

    const scope = nock(TEST_URI)
      // @todo We'd want to use `query` object here as it is much more readable, but nock for some reason
      //       fails to match request when it contains `cmd[someName]`. The issue definitely connected
      //       to the `[]`, since it does not appear when only one bracket present
      .get(`/${Method.BATCH}?cmd%5Bone%5D=${commands.one.method}%3FID%3D${dealId}`)
      .reply(RESPONSE_200, payload)
      .get(`/${Method.BATCH}?cmd%5Btwo%5D=${commands.two.method}`)
      .reply(RESPONSE_200, payload)

    await batch(commands, maxCommandsPerBatch)

    expect(scope.done()).toBe(undefined)
  })

  it('should form a proper request with more than allowed max numbered commands', async () => {
    const maxCommandsPerBatch = 1
    const payload = { result: { result: ['done1', 'done2'], result_error: [] } }
    const dealId = 999

    const commands = {
      0: { method: Method.CRM_DEAL_GET, params: { ID: dealId } },
      1: { method: Method.CRM_DEAL_LIST }
    } as const

    const scope = nock(TEST_URI)
      // @todo We'd want to use `query` object here as it is much more readable, but nock for some reason
      //       fails to match request when it contains `cmd[someName]`. The issue definitely
      //       connected to the `[]` since it does not appear when only one bracket present
      .get(`/${Method.BATCH}?cmd%5B0%5D=${commands[0].method}%3FID%3D${dealId}`)
      .reply(RESPONSE_200, payload)
      .get(`/${Method.BATCH}?cmd%5B1%5D=${commands[1].method}`)
      .reply(RESPONSE_200, payload)

    await batch(commands, maxCommandsPerBatch)

    expect(scope.done()).toBe(undefined)
  })

  it('should form a proper request with more than max allowed array of commands', async () => {
    const maxCommandsPerBatch = 1
    const payload = { result: { result: ['done1', 'done2'], result_error: [] } }
    const dealId = 999

    const commands = [
      { method: Method.CRM_DEAL_GET, params: { ID: dealId } },
      { method: Method.CRM_DEAL_LIST }
    ] as const

    const scope = nock(TEST_URI)
      // @todo We'd want to use `query` object here as it is much more readable, but nock for some reason
      //       fails to match request when it contains `cmd[someName]`. The issue definitely
      //       connected to the `[]` since it does not appear when only one bracket present
      .get(`/${Method.BATCH}?cmd%5B0%5D=${commands[0].method}%3FID%3D${dealId}`)
      .reply(RESPONSE_200, payload)
      .get(`/${Method.BATCH}?cmd%5B0%5D=${commands[1].method}`)
      .reply(RESPONSE_200, payload)

    await batch(commands, maxCommandsPerBatch)

    expect(scope.done()).toBe(undefined)
  })

  it('should return body as payload', async () => {
    const commands = {
      0: { method: Method.CRM_DEAL_GET },
      1: { method: Method.CRM_DEAL_LIST }
    }

    const payload = {
      result: {
        result: ['done'],
        result_error: [],
        result_next: [],
        result_time: [],
        result_total: []
      },
      time: {}
    }

    nock(TEST_URI)
      // @todo We'd want to use `query` object here as it is much more readable, but nock for some reason
      //       fails to match request when it contains `cmd[someName]`. The issue definitely connected
      //       to the `[]`, since it does not appear when only one bracket present
      .get(`/${Method.BATCH}?cmd%5B0%5D=${commands[0].method}&cmd%5B1%5D=${commands[1].method}`)
      .reply(RESPONSE_200, payload)

    expect(await batch(commands)).toEqual(payload)
  })

  // @todo Test this also for named and numbered commands
  it('should merge payloads when more then max allowed commands provided', async () => {
    const maxAllowedCommands = 1
    const commands = {
      0: { method: Method.CRM_DEAL_GET },
      1: { method: Method.CRM_DEAL_LIST }
    }

    const payload1 = {
      result: {
        result: ['done1'],
        result_error: [],
        result_next: [1],
        result_time: [],
        result_total: [2]
      },
      time: {}
    }

    const payload2 = {
      result: {
        result: ['done2'],
        result_error: [],
        result_next: [2],
        result_time: [],
        result_total: [2]
      },
      time: {}
    }

    nock(TEST_URI)
      // @todo We'd want to use `query` object here as it is much more readable, but nock for some reason
      //       fails to match request when it contains `cmd[someName]`. The issue definitely connected
      //       to the `[]`, since it does not appear when only one bracket present
      .get(`/${Method.BATCH}?cmd%5B0%5D=${commands[0].method}`)
      .reply(RESPONSE_200, payload1)
      .get(`/${Method.BATCH}?cmd%5B1%5D=${commands[1].method}`)
      .reply(RESPONSE_200, payload2)

    expect(await batch(commands, maxAllowedCommands)).toMatchSnapshot()
  })

  it.todo('should cast payload to the <P>')

  it('should throw when getting errors in payload', () => {
    const payload = {
      result: {
        result: { one: 'done', two: 'done' },
        result_error: {
          one: 'Expected error from numbered `batch` one',
          two: 'Expected error from numbered `batch` two'
        }
      }
    }

    const commands = {
      one: { method: Method.CRM_DEAL_GET },
      two: { method: Method.CRM_DEAL_LIST }
    }

    nock(TEST_URI)
      // @todo We'd want to use `query` object here as it is much more readable, but nock for some reason
      //       fails to match request when it contains `cmd[someName]`. The issue definitely connected
      //       to the `[]`, since it does not appear when only one bracket present
      .get(
        `/${Method.BATCH}?cmd%5Bone%5D=${commands.one.method}&cmd%5Btwo%5D=${commands.two.method}`
      )
      .reply(RESPONSE_200, payload)

    return expect(batch(commands)).rejects.toMatchSnapshot()
  })

  it('should throw when getting errors in numbered commands payload', () => {
    const payload = {
      result: {
        result: ['done'],
        result_error: [
          'Expected error from numbered `batch` 0',
          'Expected error from numbered `batch` 1'
        ]
      }
    }

    const commands = {
      0: { method: Method.CRM_DEAL_GET },
      1: { method: Method.CRM_DEAL_LIST }
    }

    nock(TEST_URI)
      // @todo We'd want to use `query` object here as it is much more readable, but nock for some reason
      //       fails to match request when it contains `cmd[someName]`. The issue definitely connected
      //       to the `[]`, since it does not appear when only one bracket present
      .get(`/${Method.BATCH}?cmd%5B0%5D=${commands[0].method}&cmd%5B1%5D=${commands[1].method}`)
      .reply(RESPONSE_200, payload)

    return expect(batch(commands)).rejects.toMatchSnapshot()
  })

  it('should throw when getting errors in array of commands payload', () => {
    const payload = {
      result: {
        result: ['done'],
        result_error: [
          'Expected error from array of command `batch` 0',
          'Expected error from array of command `batch` 1'
        ]
      }
    }

    const commands = [{ method: Method.CRM_DEAL_GET }, { method: Method.CRM_DEAL_LIST }] as const

    nock(TEST_URI)
      // @todo We'd want to use `query` object here as it is much more readable, but nock for some reason
      //       fails to match request when it contains `cmd[someName]`. The issue definitely connected
      //       to the `[]`, since it does not appear when only one bracket present
      .get(`/${Method.BATCH}?cmd%5B0%5D=${commands[0].method}&cmd%5B1%5D=${commands[1].method}`)
      .reply(RESPONSE_200, payload)

    return expect(batch(commands)).rejects.toMatchSnapshot()
  })
})
