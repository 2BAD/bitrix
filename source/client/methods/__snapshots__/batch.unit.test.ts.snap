// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Client \`batch\` method should merge payloads when more then max allowed commands provided 1`] = `
Object {
  "result": Object {
    "result": Array [
      "done1",
      "done2",
    ],
    "result_error": Array [],
    "result_next": Array [
      1,
      2,
    ],
    "result_time": Array [],
    "result_total": Array [
      2,
      2,
    ],
  },
  "time": Object {},
}
`;

exports[`Client \`batch\` method should throw when getting errors in array of commands payload 1`] = `
[Error: [batch] failed to process. Received errors in 2 commands:
Expected error from array of command \`batch\` 0
Expected error from array of command \`batch\` 1]
`;

exports[`Client \`batch\` method should throw when getting errors in numbered commands payload 1`] = `
[Error: [batch] failed to process. Received errors in 2 commands:
Expected error from numbered \`batch\` 0
Expected error from numbered \`batch\` 1]
`;

exports[`Client \`batch\` method should throw when getting errors in payload 1`] = `
[Error: [batch] failed to process. Received errors in 2 commands:
Expected error from numbered \`batch\` one
Expected error from numbered \`batch\` two]
`;

exports[`Client \`chunkCommands\` method should chunk array of commands 1`] = `
Array [
  Array [
    Object {
      "method": "crm.deal.get",
    },
    Object {
      "method": "crm.lead.get",
    },
  ],
  Array [
    Object {
      "method": "crm.deal.list",
    },
    Object {
      "method": "crm.lead.list",
    },
  ],
  Array [
    Object {
      "method": "crm.status.list",
    },
  ],
]
`;

exports[`Client \`chunkCommands\` method should chunk named commands 1`] = `
Array [
  Object {
    "a": Object {
      "method": "crm.deal.get",
    },
    "b": Object {
      "method": "crm.lead.get",
    },
  },
  Object {
    "c": Object {
      "method": "crm.deal.list",
    },
    "d": Object {
      "method": "crm.lead.list",
    },
  },
  Object {
    "e": Object {
      "method": "crm.status.list",
    },
  },
]
`;

exports[`Client \`handleBatchPayload\` method should throw when getting errors in numbered commands payload 1`] = `
"[batch] failed to process. Received errors in 1 commands:
Expected error for \`handleBatchPayload\`"
`;

exports[`Client \`handleBatchPayload\` method should throw when getting errors in payload 1`] = `
"[batch] failed to process. Received errors in 1 commands:
Expected error for \`handleBatchPayload\`"
`;

exports[`Client \`mergeBatchPayloads\` method should merge array payloads 1`] = `
Object {
  "result": Object {
    "result": Array [
      Array [
        Object {
          "ID": "1",
        },
        Object {
          "ID": "2",
        },
      ],
      Array [
        Object {
          "ID": "3",
        },
        Object {
          "ID": "4",
        },
      ],
      Array [
        Object {
          "ID": "5",
        },
        Object {
          "ID": "6",
        },
      ],
      Array [
        Object {
          "ID": "7",
        },
        Object {
          "ID": "8",
        },
      ],
    ],
    "result_error": Array [],
    "result_next": Array [
      2,
      4,
      6,
      8,
    ],
    "result_time": Array [
      Object {
        "date_finish": "2019-08-30T23:28:11+03:00",
        "date_start": "2019-08-30T23:28:11+03:00",
        "duration": 0.014085054397583008,
        "finish": 1567196891.022234,
        "processing": 0.013998985290527344,
        "start": 1567196891.008149,
      },
      Object {
        "date_finish": "2019-08-30T23:28:11+03:00",
        "date_start": "2019-08-30T23:28:11+03:00",
        "duration": 0.009933948516845703,
        "finish": 1567196891.03225,
        "processing": 0.009846210479736328,
        "start": 1567196891.022316,
      },
      Object {
        "date_finish": "2019-08-30T23:28:11+03:00",
        "date_start": "2019-08-30T23:28:11+03:00",
        "duration": 0.002981901168823242,
        "finish": 1567196891.035297,
        "processing": 0.002897024154663086,
        "start": 1567196891.032315,
      },
      Object {
        "date_finish": "2019-08-30T23:28:11+03:00",
        "date_start": "2019-08-30T23:28:11+03:00",
        "duration": 0.004575967788696289,
        "finish": 1567196891.039936,
        "processing": 0.00450897216796875,
        "start": 1567196891.03536,
      },
    ],
    "result_total": Array [
      8,
      8,
      8,
      8,
    ],
  },
  "time": Object {
    "date_finish": "2019-08-30T23:28:13+03:00",
    "date_start": "2019-08-30T23:28:12+03:00",
    "duration": 0.2647218704223683,
    "finish": 1567196891.225739,
    "processing": 0.21567082405090832,
    "start": 1567196890.959517,
  },
}
`;

exports[`Client \`mergeBatchPayloads\` method should merge named payloads 1`] = `
Object {
  "result": Object {
    "result": Object {
      "a": Array [
        Object {
          "ID": "1",
        },
        Object {
          "ID": "2",
        },
      ],
      "b": Array [
        Object {
          "ID": "3",
        },
        Object {
          "ID": "4",
        },
      ],
      "c": Array [
        Object {
          "ID": "5",
        },
        Object {
          "ID": "6",
        },
      ],
      "d": Array [
        Object {
          "ID": "7",
        },
        Object {
          "ID": "8",
        },
      ],
    },
    "result_error": Array [],
    "result_next": Object {
      "a": 2,
      "b": 4,
      "c": 6,
      "d": 8,
    },
    "result_time": Object {
      "a": Object {
        "date_finish": "2019-08-30T23:28:11+03:00",
        "date_start": "2019-08-30T23:28:11+03:00",
        "duration": 0.014085054397583008,
        "finish": 1567196891.022234,
        "processing": 0.013998985290527344,
        "start": 1567196891.008149,
      },
      "b": Object {
        "date_finish": "2019-08-30T23:28:11+03:00",
        "date_start": "2019-08-30T23:28:11+03:00",
        "duration": 0.009933948516845703,
        "finish": 1567196891.03225,
        "processing": 0.009846210479736328,
        "start": 1567196891.022316,
      },
      "c": Object {
        "date_finish": "2019-08-30T23:28:11+03:00",
        "date_start": "2019-08-30T23:28:11+03:00",
        "duration": 0.002981901168823242,
        "finish": 1567196891.035297,
        "processing": 0.002897024154663086,
        "start": 1567196891.032315,
      },
      "d": Object {
        "date_finish": "2019-08-30T23:28:11+03:00",
        "date_start": "2019-08-30T23:28:11+03:00",
        "duration": 0.004575967788696289,
        "finish": 1567196891.039936,
        "processing": 0.00450897216796875,
        "start": 1567196891.03536,
      },
    },
    "result_total": Object {
      "a": 8,
      "b": 8,
      "c": 8,
      "d": 8,
    },
  },
  "time": Object {
    "date_finish": "2019-08-30T23:28:13+03:00",
    "date_start": "2019-08-30T23:28:12+03:00",
    "duration": 0.2647218704223683,
    "finish": 1567196891.225739,
    "processing": 0.21567082405090832,
    "start": 1567196890.959517,
  },
}
`;

exports[`Client \`prepareCommandsQueries\` method should return empty query object when no commands provided 1`] = `Object {}`;

exports[`Client \`prepareCommandsQueries\` method should transform dict of the commands into the query object 1`] = `
Object {
  "cmd[one]": "crm.deal.get?ID=11111",
  "cmd[two]": "crm.deal.list?filter%5B%3EPROBABILITY%5D=50&order%5BSTAGE_ID%5D=ASC&select%5B0%5D=ID&select%5B1%5D=TITLE&start=200",
}
`;

exports[`Client \`prepareCommandsQueries\` method should work with array of commands 1`] = `
Object {
  "cmd[0]": "crm.deal.get",
  "cmd[1]": "crm.deal.list",
}
`;

exports[`Client \`prepareCommandsQueries\` method should work with numbered commands 1`] = `
Object {
  "cmd[0]": "crm.deal.get",
  "cmd[1]": "crm.deal.list",
}
`;
